### ISSUE TITLE
Critical Model Issues: Collapse and Poor Performance

### ISSUE THREAD

**Issue by Bempong-Sylvester-Obese** [Bempong-Sylvester-Obese/RSNA-Intracranial-Aneurysm-Detection#7]
Problem Summary
The current aneurysm detection model exhibits severe training and performance issues that prevent it from being clinically viable.
🔍 Issues Identified
1. Model Collapse - Critical

Symptom: Model predicts identical probability (0.369) for ALL test samples
Impact: Complete failure to distinguish between aneurysm and non-aneurysm cases
Evidence:
✅ True: 0, Pred: 0, Prob: 0.369
❌ True: 1, Pred: 0, Prob: 0.369
✅ True: 0, Pred: 0, Prob: 0.369
❌ True: 1, Pred: 0, Prob: 0.369

2. Random Performance

AUC-ROC: 0.500 (no better than random chance)
Classification: Only predicts "No Aneurysm" class
Precision/Recall: 0.00 for aneurysm detection

3. Architecture Problems

Overly complex U-Net design for classification task
Inappropriate combination of segmentation + classification losses
Poor weight initialization leading to gradient issues

4. Training Instability

Aggressive timeout handling causing data loading failures
Complex error recovery masking underlying issues
Multiple progress bars creating confusion

5. Data Processing Issues

Dummy mask generation instead of real segmentation targets
Inadequate DICOM preprocessing
Poor handling of class imbalance

📊 Current Performance Metrics
🎯 Accuracy: 0.640 (misleading - only due to class imbalance)
📈 AUC-ROC: 0.500 (random performance)
📋 Classification Report:
              precision    recall  f1-score   support
 No Aneurysm       0.64      1.00      0.78        64
    Aneurysm       0.00      0.00      0.00        36
🛠️ Root Causes Analysis
1. Inappropriate Architecture
python# PROBLEM: Complex U-Net for simple classification
class TPUSegmentationModel(nn.Module):
    def __init__(self):
        self.backbone = TPUOptimized3DUNet(...)  # Overkill
        self.seg_head = nn.Conv3d(...)          # Unnecessary
        self.classifier = nn.Sequential(...)     # Poorly connected
2. Conflicting Loss Functions
python# PROBLEM: Mixing segmentation and classification losses
seg_loss = nn.functional.binary_cross_entropy_with_logits(seg_logits, mask)
cls_loss = nn.functional.binary_cross_entropy_with_logits(cls_logits.view(-1), has_aneurysm)
total_loss = seg_loss + 0.5 * cls_loss  # Confuses training objective
3. Dummy Data Issues
python# PROBLEM: Fake segmentation masks
def _create_dummy_mask(self, shape, has_aneurysm):
    mask = np.zeros(shape, dtype=np.float32)
    if has_aneurysm:
        # Creates meaningless rectangular regions
✅ Proposed Solutions
1. Simplified Architecture

Replace U-Net with focused 3D CNN for classification
Remove segmentation components entirely
Implement proper medical imaging architecture

2. Fixed Training Pipeline
python# SOLUTION: Single objective training
class ImprovedAneurysmNet(nn.Module):
    def __init__(self):
        self.features = nn.Sequential(...)      # 3D feature extraction
        self.classifier = nn.Sequential(...)    # Direct classification
        
# SOLUTION: Balanced loss function
criterion = BalancedBCELoss(pos_weight=class_weights)
3. Better Data Processing

Robust DICOM loading without timeouts
Proper volume normalization using medical imaging best practices
Automatic class balancing

4. Stable Training Loop

Remove complex timeout/error handling
Single progress bar per epoch
Proper TPU synchronization

🎯 Expected Improvements
After implementing fixes:

AUC-ROC: > 0.75 (from 0.50)
Accuracy: > 0.80 (from 0.64)
Predictions: Diverse probabilities instead of constant 0.369
Training: Stable convergence without crashes
Classification: Both classes properly detected

📋 Implementation Checklist

 Replace TPUSegmentationModel with ImprovedAneurysmNet
 Remove segmentation-related code (masks, seg_head, seg_loss)
 Implement BalancedBCELoss with automatic class weighting
 Fix DICOM preprocessing pipeline
 Simplify training loop and remove timeout handling
 Add proper model evaluation metrics
 Test on validation set to confirm fixes

🔄 Testing Strategy

Unit Tests: Test individual components (DICOM loading, model forward pass)
Integration Tests: Full training pipeline on small dataset
Performance Tests: Validation on held-out test set
Regression Tests: Ensure fixes don't break TPU compatibility

📝 Files Affected

aneurysmnet-ipynb.ipynb - Main notebook requiring complete refactor
Model architecture classes
Training pipeline functions
Evaluation functions

⚠️ Priority: HIGH
This issue prevents the model from being used for any practical aneurysm detection. The current state is completely non-functional for medical diagnosis and requires immediate attention.
